File: cpp/tests/geometry/NeighborSearch.cpp
// ----------------------------------------------------------------------------
// -                        Open3D: www.open3d.org                            -
// ----------------------------------------------------------------------------
// The MIT License (MIT)
//
// Copyright (c) 2018 www.open3d.org
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
// IN THE SOFTWARE.
// ----------------------------------------------------------------------------
#include "open3d/geometry/NeighborSearch.h"
#include "open3d/core/Device.h"
#include "open3d/core/Dtype.h"
#include "open3d/core/SizeVector.h"
#include "open3d/core/Tensor.h"
#include "open3d/geometry/PointCloud.h"
#include "open3d/geometry/TriangleMesh.h"

#include "core/CoreTest.h"
#include "tests/UnitTest.h"

using namespace Eigen;
using namespace open3d;
using namespace std;

namespace open3d {
namespace tests {

TEST(NeighborSearch, SearchKNN) {
    vector<int> ref_indices = {27, 48, 4,  77, 90, 7,  54, 17, 76, 38,
                               39, 60, 15, 84, 11, 57, 3,  32, 99, 36,
                               52, 40, 26, 59, 22, 97, 20, 42, 73, 24};

    vector<float> ref_distance2 = {
            0.000000,  4.684353,  4.996539,  9.191849,  10.034604, 10.466745,
            10.649751, 11.434066, 12.089195, 13.345638, 13.696270, 14.016148,
            16.851978, 17.073435, 18.254518, 20.019994, 21.496347, 23.077277,
            23.692427, 23.809303, 24.104578, 25.005770, 26.952710, 27.487888,
            27.998463, 28.262975, 28.581313, 28.816608, 31.603230, 31.610916};

    int size = 100;
    float threshold = 1e-5;

    geometry::PointCloud pc;

    Vector3d vmin(0.0, 0.0, 0.0);
    Vector3d vmax(10.0, 10.0, 10.0);

    pc.points_.resize(size);
    Rand(pc.points_, vmin, vmax, 0);

    std::vector<float> points;
    points.resize(100 * 3);
    for (unsigned int i = 0; i < pc.points_.size(); i++) {
        for (unsigned int j = 0; j < 3; j++) {
            points[3 * i + j] = (float)pc.points_[i].data()[j];
        }
    }

    core::Device device;
    core::Tensor t(points, {100, 3}, core::Dtype::Float32, device);

    geometry::NeighborSearch neighborsearch(t);

    vector<float> query_vec = {1.647059, 4.392157, 8.784314};
    core::Tensor q(query_vec, {1, 3}, core::Dtype::Float32, device);
    int knn = 30;
    vector<long> indices;
    vector<float> distance2;

    pair<core::Tensor, core::Tensor> result = neighborsearch.SearchKNN(q, knn);

    indices = result.first.ToFlatVector<long>();
    distance2 = result.second.ToFlatVector<float>();
    vector<int> indices2(indices.begin(), indices.end());

    ExpectEQ(ref_indices, indices2);
    ExpectEQ(ref_distance2, distance2, threshold);
}

TEST(NeighborSearch, SearchKNN_MultiQuery) {
    vector<int> ref_indices = {27, 48, 4,  77, 90, 7,  54, 17, 76, 38,
                               39, 60, 15, 84, 11, 
                               28, 94, 92, 33, 18, 68, 44, 58, 21, 23, 
                               80, 87, 64, 35, 89};

    vector<float> ref_distance2 = {
            0.000000,  4.684353,  4.996539,  9.191849,  10.034604, 10.466745,
            10.649751, 11.434066, 12.089195, 13.345638, 13.696270, 14.016148,
            16.851978, 17.073435, 18.254518, 
            0.000000,  0.187620,  1.527110,  3.031130,  3.131100,  5.7424, 
            7.271060,  8.319890,  10.08070,  10.12840,  10.97730,  12.52132, 
            13.25340,  14.01610,  14.56210};

    int size = 100;
    float threshold = 5e-5;

    geometry::PointCloud pc;

    Vector3d vmin(0.0, 0.0, 0.0);
    Vector3d vmax(10.0, 10.0, 10.0);

    pc.points_.resize(size);
    Rand(pc.points_, vmin, vmax, 0);

    std::vector<float> points;
    points.resize(100 * 3);
    for (unsigned int i = 0; i < pc.points_.size(); i++) {
        for (unsigned int j = 0; j < 3; j++) {
            points[3 * i + j] = (float)pc.points_[i].data()[j];
        }
    }

    core::Device device;
    core::Tensor t(points, {100, 3}, core::Dtype::Float32, device);

    geometry::NeighborSearch neighborsearch(t);

    vector<float> query_vec = {1.647059, 4.392157, 8.784314,
                               8.274510, 3.294120, 2.274510};
    core::Tensor q(query_vec, {2, 3}, core::Dtype::Float32, device);
    int knn = 15;
    vector<long> indices;
    vector<float> distance2;

    pair<core::Tensor, core::Tensor> result = neighborsearch.SearchKNN(q, knn);

    indices = result.first.ToFlatVector<long>();
    distance2 = result.second.ToFlatVector<float>();
    vector<int> indices2(indices.begin(), indices.end());

    ExpectEQ(ref_indices, indices2);
    ExpectEQ(ref_distance2, distance2, threshold);
}

TEST(NeighborSearch, SearchKNN_2e6) {
    vector<int> ref_indices = {27,      1152930, 1224678, 1913298, 241116, 1006205, 1059237, 1714823, 1074643, 238765, 
                               1346254, 937272,  1766692, 401566,  544783, 1003373, 1925626, 1217426, 366060,  1054994, 
                               1808440, 494538,  1483669, 1982240, 324281, 1504383, 1767616, 41418,   446301,  646945};

    vector<float> ref_distance2 = {
            0.00000, 0.25552, 0.27129, 0.38407, 0.45310, 0.65057, 0.78246, 0.96896, 0.98150, 1.04516, 
            1.15342, 1.16118, 1.31195, 1.39012, 1.48130, 1.48707, 1.49334, 1.50074, 1.52727, 1.59777, 
            1.76207, 1.77287, 1.79251, 1.79462, 1.83828, 1.89121, 1.92284, 2.00621, 2.10253, 2.12210};

    srand(0);
    unsigned int size = 2e6;
    float threshold = 1e-5;

    std::vector<float> points;
    points.resize(2e6 * 3);
    for (unsigned int i = 0; i < size; i++) {
        for (unsigned int j = 0; j < 3; j++) {
            points[3 * i + j] = ((float)rand() / RAND_MAX) * 100.0;
        }
    }

    core::Device device;
    core::Tensor t(points, {(long)2e6, 3}, core::Dtype::Float32, device);

    geometry::NeighborSearch neighborsearch(t);

    vector<float> query_vec = {16.5974, 44.0105, 88.0075};
    core::Tensor q(query_vec, {1, 3}, core::Dtype::Float32, device);
    int knn = 30;
    vector<long> indices;
    vector<float> distance2;

    pair<core::Tensor, core::Tensor> result = neighborsearch.SearchKNN(q, knn);

    indices = result.first.ToFlatVector<long>();
    distance2 = result.second.ToFlatVector<float>();
    vector<int> indices2(indices.begin(), indices.end());

    ExpectEQ(ref_indices, indices2);
    ExpectEQ(ref_distance2, distance2, threshold);
}

TEST(NeighborSearch, SearchKNN_GPU_MultiQuery) {
    vector<int> ref_indices = {27, 48, 4,  77, 90, 7,  54, 17, 76, 38,
                               39, 60, 15, 84, 11, 
                               28, 94, 92, 33, 18, 68, 44, 58, 21, 23, 
                               80, 87, 64, 35, 89};

    vector<float> ref_distance2 = {
            0.000000,  4.684353,  4.996539,  9.191849,  10.034604, 10.466745,
            10.649751, 11.434066, 12.089195, 13.345638, 13.696270, 14.016148,
            16.851978, 17.073435, 18.254518, 
            0.000000,  0.187620,  1.527110,  3.031130,  3.131100,  5.7424, 
            7.271060,  8.319890,  10.08070,  10.12840,  10.97730,  12.52132, 
            13.25340,  14.01610,  14.56210};

    int size = 100;
    float threshold = 5e-5;

    geometry::PointCloud pc;

    Vector3d vmin(0.0, 0.0, 0.0);
    Vector3d vmax(10.0, 10.0, 10.0);

    pc.points_.resize(size);
    Rand(pc.points_, vmin, vmax, 0);

    std::vector<float> points;
    points.resize(100 * 3);
    for (unsigned int i = 0; i < pc.points_.size(); i++) {
        for (unsigned int j = 0; j < 3; j++) {
            points[3 * i + j] = (float)pc.points_[i].data()[j];
        }
    }

    core::Device device("CUDA:0");
    core::Tensor t(points, {100, 3}, core::Dtype::Float32, device);

    geometry::NeighborSearch neighborsearch(t);

    vector<float> query_vec = {1.647059, 4.392157, 8.784314,
                               8.274510, 3.294120, 2.274510};
    core::Tensor q(query_vec, {2, 3}, core::Dtype::Float32, device);
    int knn = 15;
    vector<long> indices;
    vector<float> distance2;

    pair<core::Tensor, core::Tensor> result = neighborsearch.SearchKNN(q, knn);

    indices = result.first.ToFlatVector<long>();
    distance2 = result.second.ToFlatVector<float>();
    vector<int> indices2(indices.begin(), indices.end());

    ExpectEQ(ref_indices, indices2);
    ExpectEQ(ref_distance2, distance2, threshold);
}

TEST(NeighborSearch, SearchHybrid) {
    vector<int> ref_indices = {27, 48, 4,  77, 90, 7,  54, 17, -1, -1,
                               -1, -1, -1, -1, -1, 
                               28, 94, 92, 33, 18, 68, 44, 58, 21, 23, 
                               80, -1, -1, -1, -1};

    vector<float> ref_distance2 = {
            0.000000,  4.684353,  4.996539, 9.191849, 10.034604, 10.466745,
            10.649751, 11.434066, 0.000000, 0.000000, 0.000000,  0.000000,
            0.000000,  0.000000,  0.000000, 
            0.000000,  0.187620,  1.527110,  3.031130,  3.131100,  5.7424, 
            7.271060,  8.319890,  10.08070,  10.12840,  10.97730,  0.000000, 
            0.000000,  0.000000,  0.000000};

    int size = 100;
    float threshold = 5e-4;

    geometry::PointCloud pc;

    Vector3d vmin(0.0, 0.0, 0.0);
    Vector3d vmax(10.0, 10.0, 10.0);

    pc.points_.resize(size);
    Rand(pc.points_, vmin, vmax, 0);

    std::vector<float> points;
    points.resize(100 * 3);
    for (unsigned int i = 0; i < pc.points_.size(); i++) {
        for (unsigned int j = 0; j < 3; j++) {
            points[3 * i + j] = (float)pc.points_[i].data()[j];
        }
    }

    core::Device device;
    core::Tensor t(points, {100, 3}, core::Dtype::Float32, device);

    geometry::NeighborSearch neighborsearch(t);

    vector<float> query_vec = {1.647059, 4.392157, 8.784314,
                               8.274510, 3.294120, 2.274510};
    core::Tensor q(query_vec, {2, 3}, core::Dtype::Float32, device);
    int max_nn = 15;
    float radius = 12;
    vector<long> indices;
    vector<float> distance2;

    pair<core::Tensor, core::Tensor> result = neighborsearch.SearchHybrid(q, radius, max_nn);

    indices = result.first.ToFlatVector<long>();
    distance2 = result.second.ToFlatVector<float>();
    vector<int> indices2(indices.begin(), indices.end());

    ExpectEQ(ref_indices, indices2);
    ExpectEQ(ref_distance2, distance2, threshold);
}

}  // namespace tests
}  // namespace open3d

File: python_old/open3d/ml/tf/python/layers/neighbor_search.py
# ----------------------------------------------------------------------------
# -                        Open3D: www.open3d.org                            -
# ----------------------------------------------------------------------------
# The MIT License (MIT)
#
# Copyright (c) 2020 www.open3d.org
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
# IN THE SOFTWARE.
# ----------------------------------------------------------------------------

from ...python.ops import ops
import tensorflow as tf

__all__ = ['FixedRadiusSearch', 'RadiusSearch', 'KNNSearch']


class FixedRadiusSearch(tf.keras.layers.Layer):
    """Fixed radius search for 3D point clouds.

    This layer computes the neighbors for a fixed radius on a point cloud.

    Example:

      This example shows a neighbor search that returns the indices to the
      found neighbors and the distances.::

        import tensorflow as tf
        import open3d.ml.tf as ml3d

        points = tf.random.normal([20,3])
        queries = tf.random.normal([10,3])
        radius = 0.8

        nsearch = ml3d.layers.FixedRadiusSearch(return_distances=True)
        ans = nsearch(points, queries, radius)
        # returns a tuple of neighbors_index, neighbors_row_splits, and neighbors_distance


    Arguments:

      metric: Either L1, L2 or Linf. Default is L2.

      ignore_query_point: If True the points that coincide with the center of
        the search window will be ignored. This excludes the query point if
        'queries' and 'points' are the same point cloud.

      return_distances: If True the distances for each neighbor will be returned.
        If False a zero length Tensor will be returned instead.
    """

    def __init__(self,
                 metric='L2',
                 ignore_query_point=False,
                 return_distances=False,
                 max_hash_table_size=32 * 2**20,
                 **kwargs):
        self.metric = metric
        self.ignore_query_point = ignore_query_point
        self.return_distances = return_distances
        self.max_hash_table_size = max_hash_table_size
        super().__init__(autocast=False, **kwargs)

    def build(self, inp_shape):
        super().build(inp_shape)

    def call(self,
             points,
             queries,
             radius,
             points_row_splits=None,
             queries_row_splits=None,
             hash_table_size_factor=1 / 64,
             hash_table=None):
        """This function computes the neighbors within a fixed radius for each query point.

        Arguments:

          points: The 3D positions of the input points. *This argument must be
            given as a positional argument!*

          queries: The 3D positions of the query points.

          radius: A scalar with the neighborhood radius

          points_row_splits: Optional 1D vector with the row splits information
            if points is batched. This vector is [0, num_points] if there is
            only 1 batch item.

          queries_row_splits: Optional 1D vector with the row splits information
            if queries is batched.  This vector is [0, num_queries] if there is
            only 1 batch item.

          hash_table_size_factor: Scalar. The size of the hash table as fraction
            of points.

          hash_table: A precomputed hash table generated with build_spatial_hash_table().
            This input can be used to explicitly force the reuse of a hash table in special
            cases and is usually not needed.
            Note that the hash table must have been generated with the same 'points' array.

        Returns:
          3 Tensors in the following order

          neighbors_index
            The compact list of indices of the neighbors. The corresponding query point
            can be inferred from the 'neighbor_count_row_splits' vector.

          neighbors_row_splits
            The exclusive prefix sum of the neighbor count for the query points including
            the total neighbor count as the last element. The size of this array is the
            number of queries + 1.

          neighbors_distance
            Stores the distance to each neighbor if 'return_distances' is True.
            Note that the distances are squared if metric is L2.
            This is a zero length Tensor if 'return_distances' is False.
        """
        if points_row_splits is None:
            points_row_splits = tf.cast(tf.stack([0, tf.shape(points)[0]]),
                                        dtype=tf.int64)
        if queries_row_splits is None:
            queries_row_splits = tf.cast(tf.stack([0, tf.shape(queries)[0]]),
                                         dtype=tf.int64)
        if hash_table is None:
            table = ops.build_spatial_hash_table(
                max_hash_table_size=self.max_hash_table_size,
                points=points,
                radius=radius,
                points_row_splits=points_row_splits,
                hash_table_size_factor=hash_table_size_factor)
        else:
            table = hash_table
        result = ops.fixed_radius_search(
            ignore_query_point=self.ignore_query_point,
            return_distances=self.return_distances,
            metric=self.metric,
            points=points,
            queries=queries,
            radius=radius,
            points_row_splits=points_row_splits,
            queries_row_splits=queries_row_splits,
            hash_table_splits=table.hash_table_splits,
            hash_table_index=table.hash_table_index,
            hash_table_cell_splits=table.hash_table_cell_splits)
        return result


class RadiusSearch(tf.keras.layers.Layer):
    """Radius search for 3D point clouds.

    This layer computes the neighbors for each query point with each query
    having an individual radius.

    Example:

      This example shows a neighbor search that returns the indices to the
      found neighbors and the distances.::

        import tensorflow as tf
        import open3d.ml.tf as ml3d

        points = tf.random.normal([20,3])
        queries = tf.random.normal([10,3])
        radii = tf.random.normal([10], mean=1.0)

        nsearch = ml3d.layers.RadiusSearch(return_distances=True)
        ans = nsearch(points, queries, radii)
        # returns a tuple of neighbors_index, neighbors_row_splits, and neighbors_distance


    Arguments:

      metric: Either L1, L2 or Linf. Default is L2.

      ignore_query_point: If True the points that coincide with the center of the
        search window will be ignored. This excludes the query point if 'queries'
        and 'points' are the same point cloud.

      return_distances: If True the distances for each neighbor will be returned.
        If False a zero length Tensor will be returned instead.

      normalize_distances: If True the returned distances will be normalized with
        the radii.
    """

    def __init__(self,
                 metric='L2',
                 ignore_query_point=False,
                 return_distances=False,
                 normalize_distances=False,
                 **kwargs):
        self.metric = metric
        self.ignore_query_point = ignore_query_point
        self.return_distances = return_distances
        self.normalize_distances = normalize_distances
        super().__init__(autocast=False, **kwargs)

    def build(self, inp_shape):
        super().build(inp_shape)

    def call(self,
             points,
             queries,
             radii,
             points_row_splits=None,
             queries_row_splits=None):
        """This function computes the neighbors within a radius for each query point.

        Arguments:

          points: The 3D positions of the input points. *This argument must be
            given as a positional argument!*

          queries: The 3D positions of the query points.

          radii: A radius for each query point.

          points_row_splits: Optional 1D vector with the row splits information
            if points is batched.  This vector is [0, num_points] if there is
            only 1 batch item.

          queries_row_splits: Optional 1D vector with the row splits information
            if queries is batched. This vector is [0, num_queries] if there is
            only 1 batch item.

        Returns:
          3 Tensors in the following order

          neighbors_index
            The compact list of indices of the neighbors. The corresponding query point
            can be inferred from the 'neighbor_count_row_splits' vector.

          neighbors_row_splits
            The exclusive prefix sum of the neighbor count for the query points including
            the total neighbor count as the last element. The size of this array is the
            number of queries + 1.

          neighbors_distance
            Stores the distance to each neighbor if 'return_distances' is True.
            Note that the distances are squared if metric is L2.
            This is a zero length Tensor if 'return_distances' is False.
        """
        if points_row_splits is None:
            points_row_splits = tf.cast(tf.stack([0, tf.shape(points)[0]]),
                                        dtype=tf.int64)
        if queries_row_splits is None:
            queries_row_splits = tf.cast(tf.stack([0, tf.shape(queries)[0]]),
                                         dtype=tf.int64)

        result = ops.radius_search(ignore_query_point=self.ignore_query_point,
                                   return_distances=self.return_distances,
                                   normalize_distances=self.normalize_distances,
                                   metric=self.metric,
                                   points=points,
                                   queries=queries,
                                   radii=radii,
                                   points_row_splits=points_row_splits,
                                   queries_row_splits=queries_row_splits)
        return result


class KNNSearch(tf.keras.layers.Layer):
    """KNN search for 3D point clouds.

    This layer computes the k nearest neighbors for each query point.

    Example:

      This example shows a neighbor search that returns the indices to the
      found neighbors and the distances.::

        import tensorflow as tf
        import open3d.ml.tf as ml3d

        points = tf.random.normal([20,3])
        queries = tf.random.normal([10,3])
        k = 8

        nsearch = ml3d.layers.KNNSearch(return_distances=True)
        ans = nsearch(points, queries, k)
        # returns a tuple of neighbors_index, neighbors_row_splits, and neighbors_distance
        # Since there are more than k points and we do not ignore any points we can
        # reshape the output to [num_queries, k] with
        neighbors_index = tf.reshape(ans.neighbors_index, [10,k])
        neighbors_distance = tf.reshape(ans.neighbors_distance, [10,k])


    Arguments:

      metric: Either L1, L2 or Linf. Default is L2.

      ignore_query_point: If True the points that coincide with the center of the
        search window will be ignored. This excludes the query point if 'queries'
        and 'points' are the same point cloud.

      return_distances: If True the distances for each neighbor will be returned.
        If False a zero length Tensor will be returned instead.
    """

    def __init__(self,
                 metric='L2',
                 ignore_query_point=False,
                 return_distances=False,
                 **kwargs):
        self.metric = metric
        self.ignore_query_point = ignore_query_point
        self.return_distances = return_distances
        super().__init__(autocast=False, **kwargs)

    def build(self, inp_shape):
        super().build(inp_shape)

    def call(self,
             points,
             queries,
             k,
             points_row_splits=None,
             queries_row_splits=None):
        """This function computes the k nearest neighbors for each query point.

        Arguments:

          points: The 3D positions of the input points. *This argument must be
            given as a positional argument!*

          queries: The 3D positions of the query points.

          k: The number of nearest neighbors to search.

          points_row_splits: Optional 1D vector with the row splits information
            if points is batched.
            This vector is [0, num_points] if there is only 1 batch item.

          queries_row_splits: Optional 1D vector with the row splits information
            if queries is batched.
            This vector is [0, num_queries] if there is only 1 batch item.

        Returns: 3 Tensors in the following order

          neighbors_index
            The compact list of indices of the neighbors. The corresponding query point
            can be inferred from the 'neighbor_count_row_splits' vector.

          neighbors_row_splits
            The exclusive prefix sum of the neighbor count for the query points including
            the total neighbor count as the last element. The size of this array is the
            number of queries + 1.

          neighbors_distance
            Stores the distance to each neighbor if 'return_distances' is True.
            Note that the distances are squared if metric is L2.
            This is a zero length Tensor if 'return_distances' is False.
        """
        if points_row_splits is None:
            points_row_splits = tf.cast(tf.stack([0, tf.shape(points)[0]]),
                                        dtype=tf.int64)
        if queries_row_splits is None:
            queries_row_splits = tf.cast(tf.stack([0, tf.shape(queries)[0]]),
                                         dtype=tf.int64)

        result = ops.knn_search(ignore_query_point=self.ignore_query_point,
                                return_distances=self.return_distances,
                                metric=self.metric,
                                points=points,
                                queries=queries,
                                k=k,
                                points_row_splits=points_row_splits,
                                queries_row_splits=queries_row_splits)
        return result
